Feat/coords

ponderingdemocritus opened the pull request:
Beginning of the coordinate system to estimate travel time between two points in the realmverse.

Notes:
I did not store the coords within the existing realm metadata function as some points are negative numbers, and these cannot be bitmapped (to my knowledge). I also don't think they need to be. So i added a new storage slot and a Struct called `Point` where we store the x,y coords. Also since we cannot store floating points, i have multiplied all the x,y coordinates by 10000, which is what will be stored in the State. In the library, we run the distance formula and then divide it by 10000 to get the actual distance.

ponderingdemocritus commented:
To complete this my thinking is:

New `initiate_travel` function within Combat module. This new function would take the traveling Realm and destination Realm. We would calculate the travel time between the two Realms and store this + `time_stamp`. The player will then have to wait until this timestamp is in the past in order to `initiate_combat`. The player can choose not to `initiate_combat` if they wish, and could choose to travel to another realm, which would be calculated from their current position to the next Realm.

During travel time the Army cannot be changed, and the Army cannot be redirected.

milancermak reviewed (changes requested):
The PR is still as a draft, but you tagged me for review - not sure, but I did have a short look 🤓 

There's an issue with negative coordinates (because Cairo). Maybe we could use `https://github.com/aave-starknet-project/aave-starknet-core/pull/120` once it's ready to do the math? Or translate the whole space to be positive since that won't have any effect on the distance between points. Not sure if it won't overflow somewhere on `pow` due to the `* 10000` - do you know what are the min & max coords we have?

Other than that, please remove the unused imports as per Amarna reports.

ponderingdemocritus commented on contracts/settling_game/modules/coordinates/library.cairo:
```diff
@@ -0,0 +1,50 @@
+        alloc_locals
+        # d = √((x2-x1)2 + (y2-y1)2)
+
+        let (x) = pow(Point_2.x - Point_1.x, 2)
```
do you mean `**` ?

milancermak replied:
No. Cairo supports `**` only for constants 😞 I meant `let s = Point_2.x - Point_1.x; let x = s * s`

ponderingdemocritus commented:
squaring removes negative numbers. It uses a normal coordinates system so -180,180. So we will not get any overflows

ponderingdemocritus commented:
changed to `Travel` to be a more generalized module. Also added an example function of how I am thinking we store the travel information. It makes a generic function so we can use it for any other traveling Asset.

milancermak commented:
> squaring removes negative numbers.

Oh 😖 of course it does!

![](https://media.giphy.com/media/tMPSeKEplOfK0/giphy.gif)

milancermak commented on contracts/settling_game/modules/travel/Travel.cairo:
```diff
@@ -0,0 +1,169 @@
+    end
+
+    with_attr error_message("TRAVEL: You have not arrived"):
+        assert 0 = TRUE
```
Just `assert arrived == TRUE` here, the if above can go.
